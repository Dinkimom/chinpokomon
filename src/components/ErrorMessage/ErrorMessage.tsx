import React, { memo, ReactElement } from 'react';
import { Link } from 'react-router-dom';
import './ErrorMessage.css';

export interface IErrorMessageProps {
  error: string;
  link?: ReactElement<Link>;
}

export const ErrorMessage: React.FC<IErrorMessageProps> = memo(
  ({ error, link }) => (
    <div className="error-message">
      <span>!</span>
      <p data-testid="error-text">{error}</p>
      {link}
    </div>
  )
);
