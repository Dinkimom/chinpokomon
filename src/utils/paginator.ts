export const paginator = (c: any, m: any): Array<string | number> => {
  var current = Number(c),
    last = Number(m),
    delta = 2,
    left = current - delta,
    right = current + delta + 1,
    range = [],
    rangeWithDots = [],
    l;

  for (let i = 1; i <= last; i++) {
    if (i === 1 || i === last || (i >= left && i < right)) {
      range.push(i);
    }
  }

  for (let i of range) {
    if (l) {
      if (i - l === 2) {
        rangeWithDots.push(l + 1);
      } else if (i - l !== 1) {
        rangeWithDots.push('...');
      }
    }
    rangeWithDots.push(i);
    l = i;
  }

  return rangeWithDots;
};
