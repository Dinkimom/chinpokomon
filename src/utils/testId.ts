export class TestId {
  public baseId: string;

  constructor(id: string) {
    this.baseId = id;
  }

  get id() {
    return this.toString();
  }

  private toString() {
    return `[data-testid="${this.baseId}"]`;
  }

  static generateId(id: string) {
    return `[data-testid="${id}"]`;
  }
}
