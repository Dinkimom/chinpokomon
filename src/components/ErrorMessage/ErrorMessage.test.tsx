import { render, shallow } from 'enzyme';
import React from 'react';
import { BrowserRouter, Link } from 'react-router-dom';
import { TestId } from '../../utils/testId';
import { ErrorMessage, IErrorMessageProps } from './ErrorMessage';

const errorText = new TestId('error-text');
const testLink = new TestId('test-link');

describe('<ErrorMessage /> : initial (without link)', () => {
  const props: IErrorMessageProps = {
    error: 'Some error',
  };

  const component = shallow(<ErrorMessage {...props} />);

  it('error message renders correctly', () => {
    expect(component.find(errorText.id).text()).toBe(props.error);
  });

  it("component doesn't have passed link", () => {
    expect(component.find(testLink.id)).toHaveLength(0);
  });
});

describe('<ErrorMessage /> : initial (with link)', () => {
  const props: IErrorMessageProps = {
    error: 'Some error',
    link: <Link to="/some-path" data-testid={testLink.baseId} />,
  };

  const component = render(
    <BrowserRouter>
      <ErrorMessage {...props} />
    </BrowserRouter>
  );

  it('error message renders correctly', () => {
    expect(component.find(errorText.id).text()).toBe(props.error);
  });

  it('component has passed link', () => {
    expect(component.find(testLink.id)).toHaveLength(1);
  });
});
