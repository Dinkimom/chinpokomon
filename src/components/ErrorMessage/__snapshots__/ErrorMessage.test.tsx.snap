// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorMessage /> : initial should render correctly 1`] = `ShallowWrapper {}`;
